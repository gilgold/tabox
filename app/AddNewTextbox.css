input[type="text"] {
  font-size: .85em;
  background-color: var(--input-bg-color);
  color: var(--text-color);
  padding: 4px 10px 7px 5px;
  float: left;
  width: 290px;
  border: none;
  border-bottom: 1px solid var(--primary-color);
  padding-right: 25px;
}

input[type="text"]:focus {
  outline: none;
}

label.textbox_label {
  color: var(--text-color);
  font-size: .85em;
  font-weight: normal;
  position: absolute;
  pointer-events: none;
  left: 5px;
  top: 6px;
  transition: 0.25s ease-in-out all;
}

input[type="text"]:focus~label.textbox_label, input[type="text"]:not(:placeholder-shown)~label.textbox_label {
  top: -20px;
  left: 0;
  font-size: .75em;
  color: var(--text-color);
}

.bar {
  float: left;
  position: relative;
  display: block;
  width: 290px;
}

.bar:before, .bar:after {
  content: '';
  height: 2px;
  width: 0;
  bottom: 1px;
  position: absolute;
  background: var(--primary-color);
  transition: 0.3s ease all;
  -moz-transition: 0.3s ease all;
  -webkit-transition: 0.3s ease all;
}

.bar:before {
  left: 50%;
}

.bar:after {
  right: 50%;
}

/* active state */

input[type="text"]:focus~.bar:before, input[type="text"]:focus~.bar:after {
  width: 50%;
}

/* buttons */

.btn {
  position: relative;
  margin-left: 5px;
  display: inline-block;
  padding: 0;
  overflow: hidden;
  border-width: 0;
  outline: none;
  border-radius: 2px;
  box-shadow: 0 1px 4px rgba(0, 0, 0, .6);
  background-color: var(--primary-color);
  color: #ecf0f1;
  transition: background-color .3s;
}

.btn:hover, .btn:focus {
  background-color: #21a77e;
  cursor: pointer;
}

.btn>* {
  position: relative;
}

.btn span {
  display: inline-block;
  padding: 6px 18px;
  font-size: 1em;
  font-weight: 600;
}

.btn:before {
  content: "";
  position: absolute;
  top: 50%;
  left: 50%;
  display: block;
  width: 0;
  padding-top: 0;
  border-radius: 100%;
  background-color: rgba(236, 240, 241, .3);
  transform: translate(-50%, -50%);
}

.btn:active:before {
  width: 120%;
  padding-top: 120%;
  transition: width .2s ease-out, padding-top .2s ease-out;
}

.highlighted_note {
  font-size: .65em;
  top: 3px;
  right: 7.2rem;
  color: var(--text-color);
  display: none;
  word-break: break-word;
  white-space: normal;
  width: 110px;
  vertical-align: middle;
  margin-left: 5px;
}

.highlighter {
  font-weight: bolder;
}

.clear-button {
  position: absolute;
  top: 5px;
  right: 249px;
  border: none;
  background-color: transparent;
  color: var(--text-color);
  z-index: 10;
  cursor: pointer;
  transition: opacity .2s;
}