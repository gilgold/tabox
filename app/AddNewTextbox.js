import React, { useEffect, useRef, useState } from 'react';
import { useSetRecoilState, useRecoilValue } from 'recoil';
import './AddNewTextbox.css';
import { settingsDataState } from './atoms/settingsDataState';
import { isHighlightedState } from './atoms/globalAppSettingsState';
import { rowToHighlightState } from './atoms/animationsState';
import { getCurrentTabsAndGroups } from './utils';
import { browser } from '../static/globals';
import { useSnackbar } from 'react-simple-snackbar';
import { SnackbarStyle } from './model/SnackbarTypes';


function SaveHighlightedOnlyLabel(props) {
    const isHighlighted = useRecoilValue(isHighlightedState);
    const [totalHighlighted, setTotalHighlighted] = useState(0);

    useEffect(async () => {
        const total = (await browser.tabs.query({highlighted:true})).length;
        setTotalHighlighted(total);
    }, [])

    return <span className="highlighted_note" style={{display:(isHighlighted ? 'inline-block' : 'none')}}>Save <span className="highlighter">{totalHighlighted} selected</span> tabs</span>
}

const useFocus = () => {
    const htmlElRef = useRef(null)
    const setFocus = () => {htmlElRef.current && htmlElRef.current.focus()}

    return [ htmlElRef, setFocus ]
}

function AddNewTextbox(props) {

    const [collectionName, setName] = useState("");
    const [disabled, setDisabled] = useState(false);
    const [inputRef, setInputFocus] = useFocus();
    const settingsData = useRecoilValue(settingsDataState);
    const isHighlighted = useRecoilValue(isHighlightedState);
    const setRowToHighlight = useSetRecoilState(rowToHighlightState);
    const [openSnackbar, closeSnackbar] = useSnackbar({style: SnackbarStyle.ERROR});

    useEffect(() => {
        setInputFocus();
    }, [])


    async function handleSave() {
        if (collectionName.trim() === '') {
            openSnackbar('Please enter a name for the collection', 2000);
            return;
        }
        setDisabled(true);
        const newItem = await getCurrentTabsAndGroups(collectionName, isHighlighted);
        const newSettingsData = settingsData ? [newItem, ...settingsData] : [newItem];
        setRowToHighlight(0);
        await props.updateRemoteData(newSettingsData);
        setTimeout(() => setDisabled(false), 1000);
    }

    async function _handleKeyDown(e) {
        if (e.key === 'Enter') {
            await handleSave();
        }
      }

    return <section>
                <div className="group">
                    <input
                        type="text"
                        maxLength="50"
                        placeholder=" "
                        name="new_setting_title"
                        id="new_setting_title"
                        onKeyDown={async e => await _handleKeyDown(e)}
                        onChange={e => setName(e.target.value)}
                        ref={inputRef}
                        style={{
                            float: 'left',
                            width: '290px',
                        }}
                    />
                    <span
                        className="bar"
                        style={{
                            float: 'left',
                            width: '290px',
                        }}></span>
                    <label className="textbox_label">Save {isHighlighted ? 'selected' : 'all'} tabs as...</label>
                </div>
                <button
                    id="add_new_setting"
                    disabled={disabled}
                    className="btn"
                    onClick={async () => await handleSave()}
                    >
                        <span>Add</span>
                    </button>
                <SaveHighlightedOnlyLabel />
            </section>;
};

export default AddNewTextbox;