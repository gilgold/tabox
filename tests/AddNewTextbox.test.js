import React from 'react';
import renderer from 'react-test-renderer';
import AddNewTextbox from '../app/AddNewTextbox';
import { RecoilRoot } from 'recoil';

describe('Add new collection textbox tests', () => {
  test('Add new collection textbox renders correctly - sync disabled', () => {
    const component = renderer.create(
      <RecoilRoot>
        <AddNewTextbox />
      </RecoilRoot>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});