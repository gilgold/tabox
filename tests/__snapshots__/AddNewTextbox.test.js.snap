// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add new collection textbox tests Add new collection textbox renders correctly - sync disabled 1`] = `
<section>
  <div
    className="group"
  >
    <input
      id="new_setting_title"
      maxLength="50"
      name="new_setting_title"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder=" "
      type="text"
      value=""
    />
    <span
      className="bar"
    />
    <label
      className="textbox_label"
    >
      Search or Add collections
    </label>
    <button
      className="clear-button"
      disabled={true}
      onClick={[Function]}
      style={
        Object {
          "opacity": "0",
        }
      }
    >
      <svg
        fill="currentColor"
        height="16px"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
        />
      </svg>
    </button>
  </div>
  <button
    className="btn"
    disabled={false}
    id="add_new_setting"
    onClick={[Function]}
  >
    <span>
      Add
    </span>
  </button>
  <span
    className="highlighted_note"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Save 
    <span
      className="highlighter"
    >
      0
       selected
    </span>
     tabs
  </span>
</section>
`;
